File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> mapFreq;
        int winstart = 0;
        int res = 0;
        int maxFreq = 0;

        for (int winend = 0; winend < s.size(); winend++) {
            mapFreq[s[winend]]++;

            maxFreq = max(maxFreq, mapFreq[s[winend]]);

            if (winend - winstart + 1 - maxFreq > k) {
                mapFreq[s[winstart]]--;
                winstart++;
            }
            res = max(res, winend - winstart + 1);
        }
        return res;
    }
};

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character, Integer> map = new HashMap<>();
        int maxLen = 0, winstart = 0;
        int maxFreq = 0;

        for (int winend = 0; winend < s.length(); winend++) {
            int currFreq = map.getOrDefault(s.charAt(winend), 0);
            map.put(s.charAt(winend), currFreq + 1);

            maxFreq = Math.max(maxFreq, currFreq + 1);

            if (winend - winstart + 1 - maxFreq > k) {
                int currFreqStart = map.getOrDefault(s.charAt(winstart), 0);
                map.put(s.charAt(winstart), currFreqStart - 1);
                winstart = winstart + 1;
            }
            maxLen = Math.max(maxLen, winend - winstart + 1);
        }
        return maxLen;
    }
}